intake_bluesky/mongo_embedded.py

```diff
@@ -0,0 +1,310 @@
+                                     {'start.uid': {'$regex': f'{name}.*'}}]}
+                        matches = list(
+                                catalog._db.header.find(regex_query).limit(10))
+                        if not matches:
+                            raise KeyError(name)
+                        elif len(matches) == 1:
+                            header_doc, = matches
+                        else:
+                            match_list = '\n'.join(doc['uid'] for doc in matches)
+                            raise ValueError(
+                                f"Multiple matches to partial uid {name!r}. "
+                                f"Up to 10 listed here:\n"
+                                f"{match_list}")
+                else:
+                    if N < 0:
+                        # Interpret negative N as "the Nth from last entry".
+                        query = catalog._query
+                        cursor = (catalog._db.header.find(query)
+                                  .sort('start.time', pymongo.DESCENDING)
+                                  .skip(-N - 1)
+                                  .limit(1))
+                        try:
+                            header_doc, = cursor
+                        except ValueError:
+                            raise ValueError(
+                                f"Catalog only contains {len(catalog)} "
+                                f"runs.")
+                    else:
+                        # Interpret positive N as
+                        # "most recent entry with scan_id == N".
+                        query = {'$and': [catalog._query, {'start.scan_id': N}]}
+                        cursor = (catalog._db.header.find(query)
+                                  .sort('start.time', pymongo.DESCENDING)
+                                  .limit(1))
+                        try:
+                            header_doc, = cursor
+                        except ValueError:
+                            raise KeyError(f"No run with scan_id={N}")
+                if header_doc is None:
+                    raise KeyError(name)
+                return self._doc_to_entry(header_doc['start'][0])
+
+            def __contains__(self, key):
+                # Avoid iterating through all entries.
+                try:
+                    self[key]
+                except KeyError:
+                    return False
+                else:
+                    return True
+
+        return Entries()
+
+    def _close(self):
+        self._client.close()
+
+    def __len__(self):
+        return self._db.header.count_documents({})
+
+    def search(self, query):
+        """
+        Return a new Catalog with a subset of the entries in this Catalog.
+
+        Parameters
+        ----------
+        query : dict
+            MongoDB query.
+        """
+        if query:
+            query = {'start.' + key: val for key, val in query.items()}
```

Comment:
This is a good place to use f-strings. Slightly faster, cleaner.

Reply:
updated